proptools/compare.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proptools
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// deepCompare is like reflect.deepEqual but it gives an explanation of the first difference that it finds
+func DeepCompare(a NamedItem, b NamedItem) (comparison Comparison) {
+	return deepCompareGeneric(toReflection(a), toReflection(b), 0)
+}
+
+type NamedItem struct {
+	Name string
+	Item interface{}
+}
+
+type Comparison struct {
```

Comment:
I talked to Dan and he thinks that if the number of fields is small then the thing to do is to make the fields be separate return values, and if the number of fields grows then to consider making the return be a struct at that time. I suppose that that's enough of a rule of thumb for me to be able to follow anyway, so I'll do that.